File: block_and_transaction/get_data/reference/getBlockDataset.cpp
#include<vector>
#include<string>
#include<io.h>
#include<iostream>
#include <fstream>
#include<string>
#include <ctime>
#include "json.hpp"

using namespace std;
using namespace nlohmann;

int diyicengLength = 1000;
int diercengLength = 1000;
string shuchu = "Block_queMinnerReward_cpp.csv";
string shuchu2 = "NormalTransaction_cpp.csv";


FILE * fp1;
FILE * fp2;


void dayin(string wenzi) {
	time_t now = time(0);
	char* dt = ctime(&now);
	cout << dt << " " << wenzi << endl;
}

bool isAxiaoyuB(string & a, string & b) {
	int asize = a.size();
	int bsize = b.size();
	
	if(asize < bsize)
		return true;
	else if(asize > bsize)
		return false;
	else{
		for(int i=0;i<asize;++i)
			if((int)a.at(i) < (int)b.at(i) )
				return true;
		return false;
	}
}


void duqufenxi(int diyiceng, int blockNumber) {
	
	string dir = "block/"+to_string(diyiceng*1000)+"/"+to_string(blockNumber)+".txt";
	
	json data;
	ifstream fin(dir.data());
	fin >> data;
	json & miaomi = data["result"];

	string timestamp_str = miaomi["timestamp"].get<string>();

	int timestamp = 0;     
	sscanf(timestamp_str.data(), "%x", &timestamp);   

	int minGasPrice = -1;
	int maxGasPrice = -1;
	int sumGasPrice = -1;
	int avgGasPrice = -1;

	int changdu =  miaomi["transactions"].size() ;
	
	for(int i=0; i<changdu; ++i ) {
		int thisGasPrice;
		json & zhegetx = miaomi["transactions"][i];
		sscanf(zhegetx["gasPrice"].get<string>().data(), "%x", &thisGasPrice); 
		/*
		cout << thisGasPrice << endl;
		cin >> thisGasPrice;*/

		if(i==0) {
			minGasPrice = thisGasPrice;
			maxGasPrice = thisGasPrice;
			sumGasPrice = thisGasPrice;
			continue;
		}
		
		if(thisGasPrice < minGasPrice)
			minGasPrice = thisGasPrice;
		if(thisGasPrice > maxGasPrice)
			maxGasPrice = thisGasPrice;
		sumGasPrice += thisGasPrice;
	
		string hash = zhegetx["hash"].get<string>();
		string from = zhegetx["from"].get<string>();
		string to = "";
		string creates = "";
		string value = "";
		
		if(!zhegetx["to"].is_null())
			to = zhegetx["to"].get<string>();
		if(!zhegetx["creates"].is_null())
			creates = zhegetx["creates"].get<string>();
		if(!zhegetx["value"].is_null())
			value = zhegetx["value"].get<string>();
		
		fprintf(fp2, "%d,%d,%s,%s,%s,%s,%s\n", blockNumber, timestamp, hash.data(), from.data(), to.data(), creates.data(), value.data());
	}

	if(sumGasPrice!=-1)
		avgGasPrice = sumGasPrice/changdu;

	int size;
		sscanf(miaomi["size"].get<string>().data(), "%x", &size);   
	int difficulty;
		sscanf(miaomi["totalDifficulty"].get<string>().data(), "%x", &difficulty);   

	int transactionCount = changdu;
		
	string minerAddress = miaomi["miner"].get<string>();
		
	int gasLimit;
		sscanf(miaomi["gasLimit"].get<string>().data(), "%x", &gasLimit);   
		
	int gasUsed;
		sscanf(miaomi["gasUsed"].get<string>().data(), "%x", &gasUsed);   

	fprintf(fp1, "%d,%d,%d,%d,%d,%s,%d,%d,%d,%d,%d\n", blockNumber, timestamp, size, difficulty, transactionCount, minerAddress.data(), gasLimit, gasUsed, minGasPrice, maxGasPrice, avgGasPrice);
}


int main()
{
	fp1 = fopen(shuchu.data(), "w+");
	fp2 = fopen(shuchu2.data(), "w+");
	
	fprintf(fp1, "%s",  "blockNumber, timestamp, size, difficulty, transactionCount, minerAddress, gasLimit, gasUsed, minGasPrice, maxGasPrice, avgGasPrice\n" );
	fprintf(fp2, "%s",  "blockNumber, timestamp, transactionHash, from, to, creates, value\n");

    string path = "D:\\washParity\\block";
    

    for(int diyiceng=0; diyiceng<1000; ++ diyiceng) {
    	for(int dierceng=0; dierceng<1000; ++dierceng) {
			int blockNumber = diyiceng*1000+dierceng;
			duqufenxi(diyiceng, blockNumber);
		}
		dayin(to_string(diyiceng));
    }
	return 0;
}

File: block_and_transaction/get_data/getBlockDataset.js
var fs = require("fs");
var web3 = require('web3');

var diyicengStart = 0;
var diyicengLength = 4794;
var diercengLength = 1000;
var shuchu = "Block_queMinnerReward_"+diyicengStart+"to"+(diyicengLength-1)+".csv";
var shuchu2 = "NormalTransaction_"+diyicengStart+"to"+(diyicengLength-1)+".csv";

var file1 = fs.createWriteStream(shuchu, {flag:"w"});
var file2 = fs.createWriteStream(shuchu2, {flag:"w"});
file1.write("blockNumber, timestamp, size, difficulty, transactionCount, minerAddress, minerExtra, gasLimit, gasUsed, gasPrices\n");
file2.write("blockNumber, timestamp, transactionHash, from, to, creates, value\n");

function dayin(wenzi) {
	var mydate = new Date();
	console.log(mydate.toGMTString().substr(17), wenzi);
}


function duqufenxi(diyiceng, blockNumber) {
	var dir = "J:/eth_block_unzip/"+diyiceng*1000+"/"+blockNumber+".txt";

	var data = fs.readFileSync(dir);
	var miaomi = JSON.parse(data).result;
	var timestamp = parseInt(miaomi.timestamp, 16);


	var gasPrices = [];

	for(i in miaomi.transactions) {
		gasPrices.push( parseInt(miaomi.transactions[i].gasPrice, 16) );

		var toWrite = [
			blockNumber,
			timestamp,
			miaomi.transactions[i].hash,
			miaomi.transactions[i].from,
			miaomi.transactions[i].to,
			miaomi.transactions[i].creates,
			web3.toBigNumber(miaomi.transactions[i].value).toString(10)
		];

		file2.write(toWrite.valueOf()+"\n");
	}

	
	var minerExtra;
	try {
		minerExtra = '"'+web3.toAscii(miaomi.extraData).replace(/"/g,'""')+'"';
	}
	catch(e) {
	}

	var toWrite = [
		blockNumber,
		timestamp,
		parseInt(miaomi.size, 16),
		parseInt(miaomi.totalDifficulty, 16),
		miaomi.transactions.length,
		miaomi.miner,
		minerExtra,
		parseInt(miaomi.gasLimit, 16),
		parseInt(miaomi.gasUsed, 16),
		'"'+gasPrices.valueOf()+'"',
	];
	file1.write(toWrite.valueOf()+"\n");

}


for(var diyiceng=diyicengStart; diyiceng<diyicengLength; ++diyiceng) {
	dayin(diyiceng);

	for(var dierceng=0; dierceng<diercengLength; ++dierceng) {

		var blockNumber = diyiceng*1000+dierceng;
		duqufenxi(diyiceng, blockNumber);

	}


}
